### ISSUE TITLE
Values added to `ListProperty` aren't taken into account when realizing the property

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#7472]
Given the following code:
```
class SomeTask extends DefaultTask {
    private ListProperty<String> property = project.objects.listProperty(String)

    void addSomeValue() {
        property.add("some-value")
    }

    @Input
    ListProperty<String> getProperty() {
        return property
    }
}
```

### Expected Behavior

The value from the provider contains the added element.

### Current Behavior

The snapshot fails with no value provided to the property. If you set the property to an empty list in the constructor, everything works as expected:
```
property.set(Collections.emptyList())
```

### Your Environment

 * Build scan URL:

**Comment by lacasseio** [comment]
A real-life example of this issue can be found here: https://github.com/lacasseio/vscode-gradle-plugin/blob/master/src/main/java/io/lacasse/vscode/gradle/internal/tasks/GenerateCompileCommandsFileTask.java#L65

**Comment by big-guy** [comment]
This is a change in 5.0: https://docs.gradle.org/release-nightly/release-notes.html#incubating-factory-methods-for-creating-properties

You need to change:
```
    private ListProperty<String> property = project.objects.listProperty(String)
```
to
```
    private ListProperty<String> property = project.objects.listProperty(String).empty()
```
Since the property no longer has a default value.

**Comment by lacasseio** [comment]
My bad, I didn't follow enough the latest changes to the `Property` API.

Given that the `collectors` list gets cleared when a value is set, wouldn't it be better to fail upon `add` and `addAll` on a `ListProperty` with no value? We can check if the `value` is `NO_VALUE_COLLECTOR` to assert this mutation. I went as far as running the debugger and saw that both collectors (no value and the one I added) were present and still couldn't obviously found the issue.
